### PR TITLE
chore: add finality activation height

### PR DESCRIPTION
**Pull request by RafilxTenfen**
- Add checks for finality activation height
- Pub rand commit start height at activation height or last commited height whatever is higher
- Finality vote, only vote in blocks after the finality activation height
- Update babylon to v014

### CODE REVIEW

File: itest/test_manager.go [hunk 1]
```diff
@@ -905,13 +906,13 @@ func ParseRespBTCDelToBTCDel(resp *bstypes.BTCDelegationResponse) (btcDel *bstyp
...
 		}
 
-		if len(ud.DelegatorUnbondingSigHex) > 0 {
-			delUnbondingSig, err := bbntypes.NewBIP340SignatureFromHex(ud.DelegatorUnbondingSigHex)
-			if err != nil {
-				return nil, err
-			}
-			btcDel.BtcUndelegation.DelegatorUnbondingSig = delUnbondingSig
-		}
+		// if len(ud.DelegatorSlashingSigHex.) > 0 {
```
Comment (KonradStaniec) [inline itest/test_manager.go:909]: leftover ?

**Review by KonradStaniec — APPROVED** [review]
In general looks good 👍 though lets wait for @gitferry review

File: itest/test_manager.go (hunk 1 above)
Comment (RafilxTenfen) [reply itest/test_manager.go:909]: yes, removing it 
Thanks

File: finality-provider/service/fastsync.go [hunk 2]
```diff
@@ -53,6 +58,10 @@ func (fp *FinalityProviderInstance) FastSync(startHeight, endHeight uint64) (*Fa
 			if fp.hasProcessed(b) {
 				continue
 			}
+			// check if it is allowed to send finality
+			if b.Height < activationBlkHeight {
+				continue
+			}
```
Comment (gitferry) [inline finality-provider/service/fastsync.go:64]: maybe this check can be moved out of the for loop by comparing the activation height against start height and end height?

File: finality-provider/service/fp_instance.go [hunk 3]
```diff
@@ -395,6 +389,18 @@ func (fp *FinalityProviderInstance) tryFastSync(targetBlock *types.BlockInfo) (*
 	return fp.FastSync(startHeight, targetBlock.Height)
 }
 
+func (fp *FinalityProviderInstance) fastSyncStartHeight(lastFinalizedHeight uint64) (uint64, error) {
```
Comment (gitferry) [inline finality-provider/service/fp_instance.go:392]: `fastSyncStartHeight` is called after we have known that there's finalized blocks, indicating that activation is already started, so maybe we don't need to check activation height here?

File: util/math.go [hunk 4]
```diff
@@ -0,0 +1,11 @@
+package util
+
+func MaxUint64(values ...uint64) (max uint64) {
```
Comment (Lazar955) [inline util/math.go:3]: Nit: 
1. Could be a generic max function
2. I find it to be an antipattern to have a `util` package (becomes a code dumpster):
- suggestion: `lib/math/math.go`

**Review by Lazar955 — APPROVED** [review]
Lgtm! Minor nits:

File: finality-provider/service/fastsync.go (hunk 2 above)
Comment (RafilxTenfen) [reply finality-provider/service/fastsync.go]: done at https://github.com/babylonlabs-io/finality-provider/pull/101/commits/e03994d40a2c50373df6793f63c1be78e4e0ec3f

File: finality-provider/service/fp_instance.go (hunk 3 above)
Comment (RafilxTenfen) [reply finality-provider/service/fp_instance.go:404]: fastSyncStartHeight returns the height that fast sync should start right? 
I probably can remove the check inside the fastSync than

**Comment by RafilxTenfen** [comment]
> Maybe I was missing something, but shouldn't we add a check to prevent the fp sending finality sig before activation height?

Yeap, you are right, there was another loop that sends finality, check added at https://github.com/babylonlabs-io/finality-provider/pull/101/commits/7175eba9f6a54b7fc4b1d5273ea3f09ee687eb6a

**Review by gitferry — APPROVED** [review]
Thanks for the changes! Lgtm!

File: util/math.go (hunk 4 above)
Comment (RafilxTenfen) [reply util/math.go]: agreed, moved 
generic did not worked that easy, so I kept as MaxUint64

**Pull request merged**